format/macho/macho.go

```diff
@@ -0,0 +1,868 @@
+			})
+		case LC_IDFVMLIB, LC_LOADFVMLIB:
+			d.FieldStruct("fvmlib", func(d *decode.D) {
+				offset := d.FieldU32("offset")
+				d.FieldU32("minor_version")
+				d.FieldU32("header_addr", scalar.Hex)
+				d.FieldUTF8NullFixedLen("name", int(cmdsize)-int(offset))
+			})
+		default:
+			if _, ok := loadCommands[cmd]; !ok {
+				d.SeekRel(int64((cmdsize - 8) * 8))
+				// Seek Rel so the parts are marked unknown
+			}
+		}
+		ncmdsIdx++
+	})
+}
+
+func fatParse(d *decode.D) {
+	// Go to start of the file again
+	d.SeekAbs(0)
+	var narchs uint64
+	var ofile_offsets []uint64
+	d.FieldStruct("fat_header", func(d *decode.D) {
+		d.FieldRawLen("magic", 4*8)
+		narchs = d.FieldU32("narchs")
+		narchsIdx := 0
+
+		d.FieldStructArrayLoop("archs", "arch", func() bool {
+			return narchsIdx < int(narchs)
+		}, func(d *decode.D) {
+			// parse FatArch
+			d.FieldStruct("fat_arch", func(d *decode.D) {
+				// beware cputype and cpusubtype changes from ofile header to fat header
+				cpuType := d.FieldU32("cputype", cpuTypes)
+				d.FieldU32("cpusubtype", cpuSubTypes[cpuType])
+				ofile_offsets = append(ofile_offsets, d.FieldU32("offset"))
+				d.FieldU32("size")
+				d.FieldU32("align")
+			})
+			narchsIdx++
+		})
+	})
+	nfilesIdx := 0
+	d.FieldStructArrayLoop("files", "file", func() bool {
+		return nfilesIdx < int(narchs)
+	}, func(d *decode.D) {
+		d.SeekAbs(int64(ofile_offsets[nfilesIdx]) * 8)
+		ofileDecode(d)
+		nfilesIdx++
+	})
+}
+
+func intelSubTypeHelper(f, m uint64) uint64 {
+	return f + (m << 4)
+}
+
+func parseMachHeaderFlags(d *decode.D) {
+	d.RawLen(4)
+
+	d.RawLen(2)
+	d.FieldBool("MH_APP_EXTENSION_SAFE")
+	d.FieldBool("MH_NO_HEAP_EXECUTION")
+
+	d.FieldBool("MH_HAS_TLV_DESCRIPTORS")
+	d.FieldBool("MH_DEAD_STRIPPABLE_DYLIB")
+	d.FieldBool("MH_PIE")
+	d.FieldBool("MH_NO_REEXPORTED_DYLIBS")
+
+	d.FieldBool("MH_SETUID_SAFE")
+	d.FieldBool("MH_ROOT_SAFE")
+	d.FieldBool("MH_ALLOW_STACK_EXECUTION")
+	d.FieldBool("MH_BINDS_TO_WEAK")
+
+	d.FieldBool("MH_WEAK_DEFINES")
+	d.FieldBool("MH_CANONICAL")
+	d.FieldBool("MH_SUBSECTIONS_VIA_SYMBOLS")
+	d.FieldBool("MH_ALLMODSBOUND")
+
+	d.FieldBool("MH_PREBINDABLE")
+	d.FieldBool("MH_NOFIXPREBINDING")
+	d.FieldBool("MH_NOMULTIDEFS")
+	d.FieldBool("MH_FORCE_FLAT")
+
+	d.FieldBool("MH_TWOLEVEL")
+	d.FieldBool("MH_LAZY_INIT")
+	d.FieldBool("MH_SPLIT_SEGS")
+	d.FieldBool("MH_PREBOUND")
+
+	d.FieldBool("MH_BINDATLOAD")
+	d.FieldBool("MH_DYLDLINK")
+	d.FieldBool("MH_INCRLINK")
+	d.FieldBool("MH_NOUNDEFS")
+}
+
+func parseSegmentFlags(d *decode.D) {
+	d.RawLen(28)
+	d.FieldBool("SG_PROTECTED_VERSION_1")
+	d.FieldBool("SG_NORELOC")
+	d.FieldBool("SG_FVMLIB")
+	d.FieldBool("SG_HIGHVM")
+}
+
+func parseSectionFlags(d *decode.D) {
+	d.FieldBool("S_ATTR_PURE_INSTRUCTIONS")
+	d.FieldBool("S_ATTR_NO_TOC")
+	d.FieldBool("S_ATTR_STRIP_STATIC_SYMS")
+	d.FieldBool("S_ATTR_NO_DEAD_STRIP")
+
+	d.FieldBool("S_ATTR_LIVE_SUPPORT")
+	d.FieldBool("S_ATTR_SELF_MODIFYING_CODE")
+	d.FieldBool("S_ATTR_DEBUG")
+	d.RawLen(1)
```

Comment:
Maybe replace `RawLen`:s with one `d.FieldRawLen("unused0", 14)` etc? and the next with `d.FieldRawLen("unused0", 8)`? same reasoning as for unknown cmd bits,  bits as not totally unknown. These flags  bits are unused (padding?) or reserved i guess?

Reply:
The flag bits are reserved as they do not correspond to anything as of now. I would not go with making them `FieldRawLen` as we have no reason to make them addressable as far as I see. The reason they are grouped by 4-bit groups is that I found it more reasonable to keep track of during impl. Will change this for this and other methods.

Reply:
Is the reasoning here with `FieldRawLen` to make these fields, we have skipped over, not be marked as `unknown_`?

Reply:
I see that is the case as you have mentioned the empty slots withing decoder will be marked as unknown and clutter the result. I would still go with using the keyword `reserved` for `FieldRawLen`

Reply:
Yeap "reserved" is good here i think. Maybe there is better word then "unknown" for bits not claimed by a decoder? hard to come up with good termonology.

Maybe was unclear with addressable, didn't mean making each reserved bit a separetly, i think it's good as it is now. Now someone could make reasonable queries selecting on reserved bits and a the same time it decode tree looks nice i think.